Add Photon OS support

DheerajSShetty opened the pull request:
- Support for Linux based platform PhotonOS

- Make netmgr calls to manage the Photon OS network

Signed-off-by: Dheeraj Shetty dheerajs@vmware.com

DheerajSShetty commented:
@cppforlife Added the tests for the code changes

DheerajSShetty commented:
@cppforlife Please let me know if there is anything that I need to do from my side.

luan commented on platform/net/photonos_net_manager.go:
```diff
@@ -0,0 +1,233 @@
+// Copyright © 2018 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache License 2.0
```
Could you remove this copyright assignment from here? Copyrights are assigned to the CF Foundation on the NOTICE file at the root of the repo.

DheerajSShetty replied:
Hi Luan,
May I know where should I add these CopyRights? Does this mean we should not write the VMware Copyright in this repo at all?

cppforlife commented:
> Does this mean we should not write the VMware Copyright in this repo at
all?

correct, all of this code is copyrighted by CFF (
https://github.com/cloudfoundry/bosh-agent/blob/master/NOTICE#L1)

On Fri, Mar 16, 2018 at 2:48 PM, Dheeraj <notifications@github.com> wrote:

> *@DheerajSShetty* commented on this pull request.
> ------------------------------
>
> In platform/net/photonos_net_manager.go
> <https://github.com/cloudfoundry/bosh-agent/pull/159#discussion_r175222137>
> :
>
> > @@ -0,0 +1,233 @@
> +// Copyright © 2018 VMware, Inc. All Rights Reserved.
> +// SPDX-License-Identifier: Apache License 2.0
>
> Hi Luan,
> May I know where should I add these CopyRights? Does this mean we should
> not write the VMware Copyright in this repo at all?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cloudfoundry/bosh-agent/pull/159#discussion_r175222137>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AALV90kRygxdhRgPStp6dpw-E2ZC_6aeks5tfDMRgaJpZM4SAyCy>
> .
>

DheerajSShetty commented:
I will remove the Copyright

DheerajSShetty commented:
@cppforlife @luan I have removed the CopyRights from the file

jfmyers9 commented on platform/cert/cert_manager.go:
```diff
@@ -160,6 +173,13 @@ func (c *certManager) UpdateCertificates(certs string) error {
 		return bosherr.WrapError(err, "Running command to update certificates without retries")
 	}
 
+	if strings.Compare(c.logTag, "PhotonOSCertManager") == 0 {
```
Can you please change this to `c.logTag == "PhotonOSCertManager"` since it is clearer and more performant; however, we're wondering if the `c.logTag` is the best way to detect we are on PhotonOS here -- can you please explain that decision?

DheerajSShetty replied:
Inside the update certificate function call, I found this a rather simply way to distinctly identify the platform. Also PhotonOSCertManager logtag wont be used by other Distribution.
Please correct me if I am wrong.

jaresty replied:
We'd like to propose an alternative approach:
1. Write a script that calls `/usr/bin/c_rehash` and cats all the ca certificates into the `/etc/pki/tls/certs/ca-bundle.crt` file
2. Use that script as the `updateCmdPath` argument.
3. Remove the now-unnecessary command and logTag-testing conditional since the script has taken care of everything necessary for setting up PhotonOS.

jfmyers9 replied:
How would that script be distributed? The `updateCmdPath` seems to be primarily used for binaries that are built into the OS to perform CA certificate updates.

jrussett replied:
Preferably we would like to have a PhotonOS stemcell native script that includes the ca certificate `cat` command, e.g. /usr/bin/photonos_cert_updater. This way we can maintain the old certmanager update certificate logic and simply call a PhotonOs-specific updateCmdPath. 

This script should be baked in the Photon OS stemcell.

jfmyers9 commented on platform/cert/cert_manager.go:
```diff
@@ -160,6 +173,13 @@ func (c *certManager) UpdateCertificates(certs string) error {
 	}
 
+	if strings.Compare(c.logTag, "PhotonOSCertManager") == 0 {
+		_, _, _, err = c.runner.RunCommand("cat", fmt.Sprintf("%s*.pem > /etc/pki/tls/certs/ca-bundle.crt", c.path))
```
Can we do this natively in golang instead of shelling out to `cat`?   [`io.Copy`](https://golang.org/pkg/io/#Copy) is a native way to do what you are doing here.

jfmyers9 commented on platform/net/photonos_net_manager.go:
```diff
@@ -0,0 +1,230 @@
+	return nil
+}
+
+func (net photonosNetManager) GetConfiguredNetworkInterfaces() ([]string, error) {
```
Can you write some tests for this method?

DheerajSShetty replied:
Done.

dpb587-pivotal commented:
🔗 Pivotal Tracker, [Story #155951359](https://www.pivotaltracker.com/story/show/155951359)

DheerajSShetty commented:
Certificate manager calls a script managed by PhotonOS. Added tests for the method GetConfiguredNetworkInterfaces()

jfmyers9 commented on platform/cert/cert_manager_test.go:
```diff
@@ -385,6 +385,35 @@ var _ = Describe("Certificate Management", func() {
+
+			SharedLinuxCertManagerExamples("/etc/ssl/certs", "/usr/bin/rehash_ca_certificates.sh")
+
+			It("executes update cert command", func() {
```
This test doesn't pass when we ran it locally. Can you fix this?

DheerajSShetty replied:
Fixed.

DheerajSShetty commented:
Squashed changes

DheerajSShetty commented:
I have added a script to PhotonOS which does certificate management. Also made sure the tests pass.
I have squashed changes and named it "Add PhotonOS support"
Please let me know if there is anything to be changed.

jrussett commented on platform/net/photonos_net_manager.go:
```diff
@@ -0,0 +1,230 @@
+	interfaces := []string{}
+
+	interfacesByMacAddress, err := net.detectMacAddresses()
+	if err != nil {
```
Can you add a test for an error raised by `detectMacAddresses()`? It'd be great if you can test the same for `SetupNetworking()` which indirectly also uses `detectMacAddresses`.

christarazi commented:
Closing due to inactivity. Feel free to open again.
